login-module/src/core.js

```diff
@@ -3,84 +3,74 @@ import Quiz from './quiz'
 import Enrollment from './enrollment'
 import Certificate from './certificate'
 
-function validateOptions(options, requiredOptions, defaultOptions) {
-	const opts = { ...defaultOptions, ...options }
+function validateOptions(options, requiredOptions) {
+	const opts = { ...options }
 	const hasRequiredOptions = requiredOptions.every(item => opts[item])
 	if (!hasRequiredOptions) {
 		throw new Error(`required option missing - one of ${requiredOptions.join(', ')}`)
 	}
 	return opts
 }
 
-class GraphAcademyServices {
+export default class GraphAcademyCore {
 	constructor(options = {}) {
 		const requiredOptions = ['trainingClassName', 'enrollmentUrl', 'authResult']
-		this.options = validateOptions(options, requiredOptions, {})
-		const { stage, trainingClassName, authResult } = this.options
-		this.authResult = authResult
+		const defaultOptions = {
+			stage: 'prod',
+		}
+		this.options = { ...defaultOptions, ...validateOptions(options, requiredOptions) }
+		this.webAuth = new WebAuth({
+			clientID: 'hoNo6B00ckfAoFVzPTqzgBIJHFHDnHYu',
+			domain: 'login.neo4j.com',
+			redirectUri: `${window.location.origin}/accounts/login`,
+			audience: 'neo4j://accountinfo/',
+			scope: 'read:account-info openid email profile user_metadata',
+			responseType: 'token id_token'
+		})
+		const { stage, trainingClassName } = this.options
 		this.certificate = new Certificate(trainingClassName, stage)
 		this.enrollment = new Enrollment(trainingClassName, stage)
 		this.quiz = new Quiz(trainingClassName, stage)
+
 	}
 
-	async enrollStudentInClass (firstName, lastName) {
-		return this.enrollment.enrollStudentInClass(firstName, lastName, this.getAccessToken())
+	async login() {
+		const { webAuth } = this
+		webAuth.checkSession({}, (err, result) => {
+			if (result) this.authResult = result;
+			return [err, result];
```

Comment:
I still don't understand how JavaScript knows that this return expression inside a callback is the return value of the async function.

What if the function has two callbacks?

```js
async login () {
  doSomething((err, result) => {
    return 'a'
  })
  doSomething((err, result) => {
    return 'b'
  })
}
```

Just trying to understand how it's working :thinking: